lua/cmp/utils/window.lua

```diff
@@ -131,61 +131,59 @@ end
+        vim.api.nvim_win_set_config(self.sbar_win, style)
       else
         style.noautocmd = true
-        self.scroll_win = vim.api.nvim_open_win(buffer.ensure(self.name .. 'scroll_buf'), false, style)
+        self.sbar_win = vim.api.nvim_open_win(buffer.ensure(self.name .. 'sbar_buf'), false, style)
         local highlight = self.scrollbar == '' and 'PmenuSbar' or 'CmpScrollBar'
-        vim.api.nvim_win_set_option(self.scroll_win, 'winhighlight', 'EndOfBuffer:'..highlight..',NormalFloat:'..highlight)
+        vim.api.nvim_win_set_option(self.sbar_win, 'winhighlight', 'EndOfBuffer:' .. highlight .. ',NormalFloat:' .. highlight)
       end
-
       if self.scrollbar ~= '' then
-        local replace = {}
-        for i = 1, style.height do replace[i] = self.scrollbar end
-
-        vim.api.nvim_buf_set_lines(vim.api.nvim_win_get_buf(self.scroll_win), 0, -1, true, replace)
+        vim.api.nvim_buf_set_lines(vim.api.nvim_win_get_buf(self.sbar_win), 0, -1, true, misc.rep({ self.scrollbar }, style.height))
       end
     end
 
     -- Draw the scrollbar thumb
-    info.scrollbar.zindex = (self.style.zindex and (self.style.zindex + 2) or 2)
-
+    local thumb_offset = math.floor(info.inner_height * (vim.fn.getwininfo(self.win)[1].topline / self:get_content_height()))
+    local thumb_height = math.ceil(info.inner_height * (info.inner_height / self:get_content_height()))
+
+    local style = {
+      relative = 'editor',
+      style = 'minimal',
+      width = 1,
+      height = thumb_height,
+      row = info.row + thumb_offset + (info.border_info.visible and 1 or 0),
+      col = info.col + info.width - 1, -- info.col was already added scrollbar offset.
+      zindex = (self.style.zindex and (self.style.zindex + 2) or 2),
+    }
     if self.thumb_win and vim.api.nvim_win_is_valid(self.thumb_win) then
-      vim.api.nvim_win_set_config(self.thumb_win, info.scrollbar)
+      vim.api.nvim_win_set_config(self.thumb_win, style)
     else
-      info.scrollbar.noautocmd = true
-      self.thumb_win = vim.api.nvim_open_win(buffer.ensure(self.name .. 'thumb_buf'), false, info.scrollbar)
+      style.noautocmd = true
+      self.thumb_win = vim.api.nvim_open_win(buffer.ensure(self.name .. 'thumb_buf'), false, style)
       local highlight = self.scrollbar == '' and 'PmenuThumb' or 'CmpScrollThumb'
-      vim.api.nvim_win_set_option(self.thumb_win, 'winhighlight', 'EndOfBuffer:'..highlight..',NormalFloat:'..highlight)
+      vim.api.nvim_win_set_option(self.thumb_win, 'winhighlight', 'EndOfBuffer:'  .. highlight .. ',NormalFloat:'  .. highlight)
     end
-
     if self.scrollbar ~= '' then
-      local replace = {}
-      for i = 1, info.scrollbar.height do replace[i] = self.scrollbar end
-
-      vim.api.nvim_buf_set_lines(vim.api.nvim_win_get_buf(self.thumb_win), 0, -1, true, replace)
+      vim.api.nvim_buf_set_lines(vim.api.nvim_win_get_buf(self.sbar_win), 0, -1, true, misc.rep({ self.scrollbar }, style.height))
```

Comment:
I am using the dev-branch since I really like the completion menu border feature. This line broke something for me, since now it errors almost on every keystroke. It's because `self.sbar_win` can be nil, and `vim.api.nvim_win_get_buf` expects a number. Don't really understand everything that happens here, but judging from the diff, should it be `self.thumb_win` here instead of `self.sbar_win`? It looks like it fixes it for me just testing really fast.

Reply:
I just recreated the problem with a minimal config, and there is definitely an issue here. It looks like the issue is appearing when the documentation window needs to have a scrollbar. If I change it to `self.thumb_win` the issue disappears. See the videos I attached.

The following video is how it is right now. Notice how it doesn't error when it doesn't need to draw a scrollbar, but when it doesn't have enough space to show the entire documentation, an error appears.

https://user-images.githubusercontent.com/5160701/157007770-0007dd35-8bc7-41dc-af44-6eeb797e56a6.mov

The following video is how it is when I change `self.sbar_win` to `self.thumb_win` on line 181. Now it doesn't error and works just as expected

https://user-images.githubusercontent.com/5160701/157007858-3d800681-8869-47a7-b949-8b8b2e93d6f0.mov

I think one need to have a language server running so that there is a documentation window. If you want to try, simply install the lua sumneko language server, and use this minimal config.

```vim
if has('vim_starting')
  set encoding=utf-8
endif
scriptencoding utf-8

if &compatible
  set nocompatible
endif

let s:plug_dir = expand('/tmp/plugged/vim-plug')
if !filereadable(s:plug_dir .. '/plug.vim')
  execute printf('!curl -fLo %s/autoload/plug.vim --create-dirs https://raw.githubusercontent.com/junegunn/vim-plug/master/plug.vim', s:plug_dir)
end

execute 'set runtimepath+=' . s:plug_dir
call plug#begin(s:plug_dir)
Plug 'hrsh7th/nvim-cmp', { 'branch': 'dev' }
Plug 'hrsh7th/cmp-buffer'
Plug 'hrsh7th/cmp-nvim-lsp'
Plug 'hrsh7th/vim-vsnip'
Plug 'neovim/nvim-lspconfig'
call plug#end()
PlugInstall | quit

" Setup global configuration. More on configuration below.
lua << EOF
local cmp = require "cmp"
cmp.setup {
  snippet = {
    expand = function(args)
      vim.fn["vsnip#anonymous"](args.body)
    end,
  },

  mapping = {
    ['<CR>'] = cmp.mapping.confirm({ select = true }),
    ['<C-d>'] = cmp.mapping(cmp.mapping.scroll_docs(4), { 'i', 'c' }),                                                                                         
    ['<C-u>'] = cmp.mapping(cmp.mapping.scroll_docs(-4), { 'i', 'c' }),                                                                                        
  },                                                                                                                                                           
                                                                                                                                                               
  window = {                                                                                                                                                   
    documentation = {                                                                                                                                          
      border = 'rounded',                                                                                                                                      
      scrollbar = '║',                                                                                                                                         
    },                                                                                                                                                         
                                                                                                                                                               
    completion = {                                                                                                                                             
      border = 'rounded',                                                                                                                                      
      scrollbar = '║',                                                                                                                                         
    },                                                                                                                                                         
  },                                                                                                                                                           
                                                                                                                                                               
  sources = cmp.config.sources({                                                                                                                               
    { name = "nvim_lsp" },                                                                                                                                     
    { name = "buffer" },                                                                                                                                       
  }),                                                                                                                                                          
}                                                                                                                                                              
                                                                                                                                                               
require('lspconfig').sumneko_lua.setup({})                                                                                                                     
EOF
```

@hrsh7th

Reply:
I think it's fixed. Thank you!